lib/Transforms/Instrumentation/ShadowMemSeaDsa.cc

```diff
@@ -877,6 +911,221 @@ void ShadowDsaImpl::updateReadMod(Function &F, NodeSet &readSet,
+        worklist.push_back(v);
+
+      continue;
+    }
+
+    if (auto *gamma = dyn_cast<SelectInst>(stripped)) {
+      worklist.push_back(gamma->getFalseValue());
+      worklist.push_back(gamma->getTrueValue());
+      continue;
+    }
+
+    LOG("shadow_optimizer",
+        llvm::errs() << "Cannot retrieve all alloc sites for " << ptr
+                     << "\n\tbecause of the instruction: " << *stripped
+                     << "\n");
+    return (cache[strippedInit] = llvm::None);
+  }
+
+  return (cache[strippedInit] = allocSites);
+}
+
+bool ShadowDsaImpl::mayClobber(CallInst &memDef, CallInst &memUse, AllocSitesCache &cache) {
+  LOG("shadow_optimizer", llvm::errs() << "\n~~~~\nmayClobber(" << memDef
+                                       << ", " << memUse << ")?\n");
+
+  // Right now memUse must be a load instruction.
+  if (isMemInit(memDef))
+    return true;
+
+  // Get the corresponding pointers and check if these two pointers have
+  // disjoint allocation sites -- if not, they cannot alias.
+  // This can be extended further with offset tracking.
+
+  auto lit = m_shadowLoadToLoad.find(&memUse);
+  if (lit == m_shadowLoadToLoad.end())
+    return true;
+
+  auto getTypeSize = [this] (Type *ty) {
+    unsigned bits = m_dl->getTypeSizeInBits(ty);
+    return bits < 8 ? 1 : bits / 8;
+  };
+
+  LoadInst *concreteLoad = lit->second;
+  assert(concreteLoad);
+  assert(memUse.getNextNode() == concreteLoad);
+  Value *loadSrc = concreteLoad->getPointerOperand();
+  const unsigned loadedBytes = getTypeSize(concreteLoad->getType());
+
+  auto sit = m_shadowStoreToStore.find(&memDef);
+  if (sit == m_shadowStoreToStore.end())
+    return true;
+
+  StoreInst *concreteStore = sit->second;
+  assert(concreteStore);
+  assert(memDef.getNextNode() == concreteStore);
+  Value *storeDst = concreteStore->getPointerOperand();
+  const unsigned storedBytes =
+      getTypeSize(concreteStore->getValueOperand()->getType());
+
+  LOG("shadow_optimizer",
+      llvm::errs() << "alias(" << *loadSrc << ", " << *storeDst << ")?\n");
+
+  // If the offsets don't overlap, no clobbering may happen.
+  if (m_graph->hasCell(*loadSrc) && m_graph->hasCell(*storeDst)) {
+    const dsa::Cell &loadCell = m_graph->getCell(*loadSrc);
+    const dsa::Cell &storeCell = m_graph->getCell(*storeDst);
+    const unsigned loadStartOffset = loadCell.getOffset();
+    const unsigned loadEndOffset = loadStartOffset + loadedBytes;
+    const unsigned storeStartOffset = storeCell.getOffset();
+    const unsigned storeEndOffset = storeStartOffset + storedBytes;
+
+    LOG("shadow_optimizer",
+        llvm::errs() << "\tmayClobber[3]\n\tloadStart: " << loadStartOffset
+                     << ", loadEnd: " << loadEndOffset
+                     << "\n\tstoreStart: " << storeStartOffset
+                     << ", storeEnd: " << storeEndOffset << "\n");
+
+    if (loadEndOffset <= storeStartOffset || loadStartOffset >= storeEndOffset)
```

Comment:
If `loadCell` or `storeCell` points to an array node then I think we need to be careful. `getOffset()` will return the offset adjusted to the type of node. For instance, given a non-array cell with offset 8 and an array cell with adjusted offset 0 and let's say size 4, this code will say they cannot overlap but they can actually overlap.

Reply:
makes sense, I'll fix this

thanks!